yt/yt/server/query_tracker/yql_engine.cpp

```diff
@@ -141,11 +138,84 @@ class TYqlQueryHandler
+    TString YqlServiceName_;
+
     IChannelPtr YqlServiceChannel_;
     TPeriodicExecutorPtr ProgressGetterExecutor_;
+    TPeriodicExecutorPtr StarterExecutor_;
 
     TFuture<TTypedClientResponse<TRspStartQuery>::TResult> AsyncQueryResult_;
 
+    YT_DECLARE_SPIN_LOCK(NThreading::TSpinLock, QueryStateSpinLock_);
+    EYqlQueryState QueryState_;
+
+    void TryStart()
+    {
+        YT_LOG_DEBUG("Start YQL query attempt (Stage: %v)", Stage_);
+        auto yqlServiceChannel = WaitForFast(YqlAgentChannelProvider_->GetChannel(YqlServiceName_))
+            .ValueOrThrow();
+
+        {
+            TYqlServiceProxy proxy(yqlServiceChannel);
+            auto req = proxy.GetYqlAgentState();
+            auto rsp = WaitFor(req->Invoke()).ValueOrThrow();
+            if (!rsp->ready_to_get_queries()) {
+                YT_LOG_DEBUG("Start YQL query attempt failed, YQL agent is overloaded (Channel: %v)",
+                    yqlServiceChannel->GetEndpointDescription());
+                return;
+            }
+        }
+
+        TYqlServiceProxy::TReqStartQueryPtr startQueryReq;
+        {
+            // TODO(max42, gritukan): Implement long polling for YQL queries.
+            auto yqlServiceChannelWithBigTimeout = CreateDefaultTimeoutChannel(yqlServiceChannel, TDuration::Days(1));
+
+            TYqlServiceProxy proxy(yqlServiceChannelWithBigTimeout);
+            startQueryReq = proxy.StartQuery();
+            SetAuthenticationIdentity(startQueryReq, TAuthenticationIdentity(User_));
+            auto* yqlRequest = startQueryReq->mutable_yql_request();
+            startQueryReq->set_row_count_limit(Config_->RowCountLimit);
+            ToProto(startQueryReq->mutable_query_id(), QueryId_);
+            yqlRequest->set_query(Query_);
+            yqlRequest->set_settings(ConvertToYsonString(SettingsNode_).ToString());
+            yqlRequest->set_mode(ToProto<int>(ExecuteMode_));
+
+            for (const auto& file : Files_) {
+                auto* protoFile = yqlRequest->add_files();
+                protoFile->set_name(file->Name);
+                protoFile->set_content(file->Content);
+                protoFile->set_type(static_cast<TYqlQueryFile_EContentType>(file->Type));
+            }
+            startQueryReq->set_build_rowsets(true);
+        }
+
+        {
+            auto guard = Guard(QueryStateSpinLock_);
+            if (QueryState_ != EYqlQueryState::Pending) {
+                YT_LOG_DEBUG("Start YQL query attempt failed, query is not in pending state (State: %v)", QueryState_);
+                return;
+            }
+
+            YT_LOG_DEBUG("Start YQL query (Stage: %v, Channel: %v)",
+                Stage_,
+                yqlServiceChannel->GetEndpointDescription());
+
+            QueryState_ = EYqlQueryState::Running;
+
+            AsyncQueryResult_ = startQueryReq->Invoke();
+            AsyncQueryResult_.Subscribe(BIND(&TYqlQueryHandler::OnYqlResponse, MakeWeak(this)).Via(GetCurrentInvoker()));
```

Comment:
So we don't actually limit amount of running YQL queries to `MaxSimultaneousQueries`

There could be a delay between `rsp->ready_to_get_queries()` and increasing `ActiveQueries_` in yqlAgent in `startQueryReq->Invoke()` (For example, if QT -> YQLAgents network stops for a moment). Then many queries will think that yqlAgent is ready and start themselves, overloading yqlAgents

To prevent it I suggest introducing `Throttled` error in `YqlResponse` and handling it in QueryTracker. Probably a new `EYqlQueryState::SentToYqlAgent` state will be needed for it(or making sync call to yqlAgents)

For a first version, however, current approach is fine

Reply:
Thank you for the suggestion to return Throttling errors from the YQL agent.

However, I'm not sure that it is necessary to introduce a new "SentToYqlAgent" state or make a sync call to implement this. New states requires changes in proxies and sdk - I would like to avoid this.

Instead of introducing a new state, we could simply use a combination of the "Running" and "Pending" states. When we attempt to run a query, we would move it to the "Running" state, and if there is a throttling failure, we would revert it to "Pending".

Reply:
>Instead of introducing a new state, we could simply use a combination of the "Running" and "Pending" states. When we attempt to run a query, we would move it to the "Running" state, and if there is a throttling failure, we would revert it to "Pending".

👍 . I see you introduced "Throttled" state too. Why?

Reply:
I didn't introduce it. What do you mean?